vllm/entrypoints/openai/api_server.py

```diff
@@ -184,6 +186,28 @@ async def validation_exception_handler(_, exc):
         return JSONResponse(err.model_dump(),
                             status_code=HTTPStatus.BAD_REQUEST)
 
+    @app.exception_handler(RuntimeError)
+    async def runtime_error_handler(_, __):
+        """On generic runtime error, check to see if the engine has died.
+        It probably has, in which case the server will no longer be able to
+        handle requests. Trigger a graceful shutdown with a SIGTERM."""
+        if engine.errored and not engine.is_running:
+            logger.fatal("AsyncLLMEngine has failed, terminating server "
+                         "process")
+            os.kill(os.getpid(), signal.SIGTERM)
+
+        return Response(status_code=HTTPStatus.INTERNAL_SERVER_ERROR)
+
+    @app.exception_handler(AsyncEngineDeadError)
+    async def engine_dead_handler(_, __):
+        """Kill the server if the async engine is already dead. It will
+        not handle any further requests."""
+        logger.fatal("AsyncLLMEngine is already dead, terminating server "
+                     "process")
+        os.kill(os.getpid(), signal.SIGTERM)
+
+        return Response(status_code=HTTPStatus.INTERNAL_SERVER_ERROR)
+
```

Comment:
I would recommend gating this behavior behind an env var/argument, since automatic process death may be unexpected.

Why not just raise a normal exception, too? calling `os.kill` on self seems like an anti-pattern.

Reply:
+1 I can gate that for sure. But I do think it should be the default behavior for running the server, I don't really see the utility in leaving it up and unresponsive if we know the engine is dead.

> Why not just raise a normal exception, too? 

In this case raising an exception isn't gonna kill the server, an unhandled exception in an endpoint handler will just cause the server to respond to the request with an empty 500 (which is also what I'm explicitly doing here). The main process is waiting on the server to exit, so the server has to be notified to exit somehow. AFAIK the canonical way of doing this when using `uvicorn.run(...)` like we are is to use a termination signal. I agree that looks funky here though, It might be possible instead to do something like
```
server = uvicorn.Server(...)
server.run()
```
in our `run_server` and then use `server.shutdown()` in these handlers to try to initiate a shutdown that will then cause the main process to return from `server.run()` and exit. WDYT?

Reply:
I think that sounds cleaner, but I don't have much experience with uvicorn, so take that with a grain of salt :)

Reply:
Okay, I think I'm happy with the server shutdown code now, and I added a flag to gate this logic.

I'm noodling on some tests and going down the route of running `run_server()` in a thread with the llm engine patched to raise an error on each request, but it is looking a bit funky. Any other suggestions?

Reply:
I think it's ok for tests

Reply:
Another approach that could work would be to schedule a period task that checks the engine health and stops the http server task when the engine is unhealthy. See https://github.com/vllm-project/vllm/pull/6740

Reply:
@dtrifiro true, though I think in this case we can avoid a lot of unnecessary checking since we catch the errors as they happen